pkg/volumes/openstack/volumes.go

```diff
@@ -0,0 +1,337 @@
+		if stack.instanceName == "" {
+			return fmt.Errorf("instance name metadata was empty")
+		}
+		glog.Infof("Found instanceName=%q", stack.instanceName)
+	}
+
+	// Internal IP
+	{
+
+		server, err := servers.Get(stack.computeClient, strings.TrimSpace(stack.meta.ServerID)).Extract()
+		if err != nil {
+			return fmt.Errorf("Failed to retrieve server information from cloud: %v", err)
+		}
+		ip, err := GetServerFixedIP(server)
+		if err != nil {
+			return fmt.Errorf("error querying InternalIP from name: %v", err)
+		}
+		stack.internalIP = net.ParseIP(ip)
+		glog.Infof("Found internalIP=%q", stack.internalIP)
+	}
+
+	return nil
+}
+
+func (stack *OpenstackVolumes) MyIP() (string, error) {
+	if stack.internalIP == nil {
+		return "", fmt.Errorf("unable to determine local IP")
+	}
+	return stack.internalIP.String(), nil
+}
+
+func (stack *OpenstackVolumes) buildOpenstackVolume(d *cinderv2.Volume) (*volumes.Volume, error) {
+	etcdName := d.Name
+
+	if plainText, ok := d.Metadata[stack.nameTag]; ok {
+		tokens := strings.SplitN(plainText, "/", 2)
+		etcdName = stack.clusterName + "-" + tokens[0]
+	}
+
+	vol := &volumes.Volume{
+		ProviderID: stack.meta.ServerID,
+		MountName:  fmt.Sprintf("master-%s", d.Name),
+		EtcdName:   etcdName,
+		Info: volumes.VolumeInfo{
+			Description: d.Description,
+		},
+		Status: d.Status,
+	}
+
+	for _, attachedTo := range d.Attachments {
+		vol.AttachedTo = attachedTo.HostName
+		if attachedTo.ServerID == stack.meta.ServerID {
+			vol.LocalDevice = attachedTo.Device
+		}
+	}
+
+	return vol, nil
+}
+
+func (stack *OpenstackVolumes) matchesTags(d *cinderv2.Volume) bool {
+	for _, k := range stack.matchTagKeys {
+		_, found := d.Metadata[k]
+		if !found {
+			return false
+		}
+	}
+
+	for k, v := range stack.matchTags {
+		a, found := d.Metadata[k]
+		if !found || a != v {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (stack *OpenstackVolumes) FindVolumes() ([]*volumes.Volume, error) {
+	var volumes []*volumes.Volume
+
+	glog.V(2).Infof("Listing Openstack disks in %s/%s", stack.project, stack.meta.AvailabilityZone)
+
+	pages, err := cinderv2.List(stack.volumeClient, cinderv2.ListOpts{
+		TenantID: stack.project,
+	}).AllPages()
+	if err != nil {
+		return volumes, fmt.Errorf("FindVolumes: Failed to list volumes: %v", err)
+	}
+	vols, err := cinderv2.ExtractVolumes(pages)
+	if err != nil {
+		return volumes, fmt.Errorf("FindVolumes: Failed to extract volumes: %v", err)
+	}
+
+	for _, volume := range vols {
+		if !stack.matchesTags(&volume) {
+			continue
+		}
+		vol, err := stack.buildOpenstackVolume(&volume)
+		if err != nil {
+			glog.Warningf("skipping volume %s: %v", volume.Name, err)
+			continue
+		}
+		volumes = append(volumes, vol)
+	}
+
+	return volumes, nil
+}
+
+// FindMountedVolume implements Volumes::FindMountedVolume
+func (_ *OpenstackVolumes) FindMountedVolume(volume *volumes.Volume) (string, error) {
+	device := volume.LocalDevice
+
+	_, err := os.Stat(volumes.PathFor(device))
```

Comment:
I am little worried about this. There is "known bug" in openstack that openstack api device path does not match the path used in instance. See issue https://github.com/kubernetes/kubernetes/issues/33128 and pr https://github.com/kubernetes/kubernetes/pull/33270 I am thinking should we use also `/dev/disk/by-id` path instead of just `/dev/vdx`for instance. There is small probability that volume attachment will fail because of this